terraform/modules/spack/gitlab_object_stores.tf

```diff
@@ -0,0 +1,189 @@
+          "s3:ListBucket",
+          "s3:AbortMultipartUpload",
+          "s3:ListMultipartUploadParts",
+          "s3:ListBucketMultipartUploads"
+        ],
+        "Resource" : [for bucket in aws_s3_bucket.gitlab_object_stores : bucket.arn]
+      }
+    ]
+  })
+}
+
+resource "aws_iam_role" "gitlab_object_stores" {
+  name        = "GitlabS3Role-${var.deployment_name}"
+  description = "Managed by Terraform. Role for GitLab to assume so that it can access relevant S3 buckets."
+  assume_role_policy = jsonencode({
+    "Version" : "2012-10-17",
+    "Statement" : [
+      {
+        "Effect" : "Allow",
+        "Principal" : {
+          "Federated" : module.eks.oidc_provider_arn
+        },
+        "Action" : "sts:AssumeRoleWithWebIdentity",
+        "Condition" : {
+          "StringEquals" : {
+            "${module.eks.oidc_provider}:aud" : "sts.amazonaws.com"
+          }
+        }
+      }
+    ]
+  })
+}
+
+resource "aws_iam_role_policy_attachment" "gitlab_object_stores" {
+  role       = aws_iam_role.gitlab_object_stores.name
+  policy_arn = aws_iam_policy.gitlab_object_stores.arn
+}
+
+
+locals {
+  connection_secret_name = "gitlab-s3-bucket-secrets"
+  connection_secret_key = "connection"
+
+  backups_secret_name = "gitlab-s3-backup-bucket-secrets"
+  backups_secret_key = "config"
+}
+
+resource "kubectl_manifest" "gitlab_object_stores_config_map" {
+  yaml_body = <<-YAML
+    apiVersion: v1
+    kind: ConfigMap
+    metadata:
+      name: gitlab-s3-bucket-config
+      namespace: gitlab
+    data:
+      values.yaml: |
+        global:
+          serviceAccount:
+            create: true
+            enabled: true
+            annotations:
+              eks.amazonaws.com/role-arn: ${aws_iam_role.gitlab_object_stores.arn}
+          # https://docs.gitlab.com/charts/advanced/external-object-storage/#lfs-artifacts-uploads-packages-external-diffs-terraform-state-dependency-proxy-secure-files
+          appConfig:
+            artifacts:
+              bucket: ${aws_s3_bucket.gitlab_object_stores["artifacts"].id}
+              connection:
+                secret: ${local.connection_secret_name}
+                key: ${local.connection_secret_key}
+            uploads:
+              bucket: ${aws_s3_bucket.gitlab_object_stores["uploads"].id}
+              connection:
+                secret: ${local.connection_secret_name}
+                key: ${local.connection_secret_key}
+            backups:
+              objectStorage:
+                backend: s3
+              bucket: ${aws_s3_bucket.gitlab_object_stores["backups"].id}
+              tmpBucket: ${aws_s3_bucket.gitlab_object_stores["tmp-storage"].id}
+        # https://docs.gitlab.com/charts/advanced/external-object-storage/#backups
+        gitlab:
+          toolbox:
+            backups:
+              objectStorage:
+                config:
+                  secret: ${local.backups_secret_name}
+                  key: ${local.backups_secret_key}
```

Comment:
We don't use gitlab's backup service currently (although we should look into it i think), but gitlab still requires buckets for them.